models/DataObject/ClassDefinition/Dao.php

```diff
@@ -162,13 +163,33 @@ public function update(): void
         // add non existing columns in the table
         foreach ($this->model->getFieldDefinitions() as $key => $value) {
             if ($value instanceof DataObject\ClassDefinition\Data\ResourcePersistenceAwareInterface
-                && $value instanceof DataObject\ClassDefinition\Data) {
+                && $value instanceof DataObject\ClassDefinition\Data
+            ) {
                 // if a datafield requires more than one column in the datastore table => only for non-relation types
                 if (!$value->isRelationType()) {
                     if (is_array($value->getColumnType())) {
                         foreach ($value->getColumnType() as $fkey => $fvalue) {
                             $this->addModifyColumn($objectDatastoreTable, $key . '__' . $fkey, $fvalue, '', 'NULL');
                             $protectedDatastoreColumns[] = $key . '__' . $fkey;
+
+                            if (($value instanceof DataObject\ClassDefinition\Data\QuantityValue
+                                    || $value instanceof DataObject\ClassDefinition\Data\QuantityValueRange)
+                                && $fkey === 'unit'
+                            ) {
+                                try {
+                                    $this->db->executeQuery(
+                                        sprintf(
+                                            'ALTER TABLE `%s` ADD CONSTRAINT `%s` FOREIGN KEY (`%s`)
+                                                REFERENCES `quantityvalue_units` (`id`) ON DELETE SET NULL',
+                                            $objectDatastoreTable,
+                                            self::getForeignKeyName($objectDatastoreTable, $key . '__' . $fkey),
+                                            $key . '__' . $fkey
+                                        )
+                                    );
+                                } catch (DriverException $e) {
+                                    // Ignore if the foreign key already exists
+                                }
+                            }
```

Comment:
Any chance we could somehow extract that into a method? It seems to be same code over and over again.

WE introduced a `indexExists` method [here](https://github.com/pimcore/pimcore/pull/16871/files#diff-38cc27289cd8997402eaa2b3921c32292050c33ff9ad2f59c5606c079bb35484R188), instead of just executing the statement and ignoring the error. Maybe we could do the same for foreign keys?

Reply:
If you tell me where to abstract the code, I'm happy to do it. I did not find the right place in my opinion where we could place it...

I've added a `foreignKeyExists` and `foreignKeyDoesNotExist` function and removed the try catch